src/oci/oci_runtime.rs

```diff
@@ -0,0 +1,675 @@
+};
+use strum::AsRefStr;
+use tokio::process::Command;
+
+#[derive(Builder, Debug, Getters, Setters)]
+#[builder(pattern = "owned", setter(into))]
+// oci_runtime is the main structure to be used when interacting with the container runtime.
+pub struct OCIRuntime {
+    #[getset(get, set)]
+    #[builder(private, default = "Box::new(DefaultOCIRuntimeExecCommand)")]
+    /// The executor for the OCIRuntime
+    exec: Box<dyn ExecCommand>,
+
+    #[get]
+    /// Path to the oci_runtime binary
+    binary: PathBuf,
+}
+
+impl OCIRuntime {
+    /// Run OCIRuntime with the provided subcommand and args and return the output if the command execution succeeds.
+    /// This can still mean that oci_runtime itself failed, which can be verified via the exist status
+    /// of the output.
+    pub async fn run(&self, subcommand: &Subcommand, args: &[GlobalArgs]) -> Result<Output> {
+        self.exec()
+            .run_output(self.binary(), &subcommand.build_cmd()[..], args)
+            .await
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+/// Defaultoci_runtimeExecCommand is a wrapper which can be used to execute oci_runtime in a standard way.
+struct DefaultOCIRuntimeExecCommand;
+
+impl ExecCommand for DefaultOCIRuntimeExecCommand {}
+
+#[async_trait]
+trait ExecCommand: Debug + DynClone + Send + Sync {
+    /// Run a command and return its `Output`.
+    async fn run_output(
+        &self,
+        binary: &Path,
+        cmd: &[String],
+        global_args: &[GlobalArgs],
+    ) -> Result<Output> {
+        Command::new(binary)
+            .args(cmd)
+            .args(global_args.iter().map(ToString::to_string))
+            .output()
+            .await
+            .context("run OCIRuntime")
+    }
+}
+
+clone_trait_object!(ExecCommand);
+
+#[derive(AsRefStr, Clone, Debug, Hash, Eq, PartialEq)]
+#[strum(serialize_all = "kebab_case")]
+pub enum Subcommand {
+    Checkpoint((String, Vec<CheckpointArgs>)),
+    Create((String, Vec<CreateArgs>)),
+    Delete(String),
+    Events((String, Vec<EventsArgs>)),
+    Exec((String, Vec<ExecArgs>)),
+    Init(),
+    Kill((String, Vec<KillArgs>)),
+    List(Vec<ListArgs>),
+    Pause(String),
+    Ps((String, Vec<PsArgs>)),
+    Restore((String, Vec<RestoreArgs>)),
+    Resume(String),
+    Run((String, Vec<RunArgs>)),
+    Spec(Vec<SpecArgs>),
+    Start(String),
+    State(String),
+    Update((String, Vec<UpdateArgs>)),
+}
+
+/// Build a vec of `[command][args][container_id]`
+fn build_vec(command: String, args: Vec<String>, container_id: Option<String>) -> Vec<String> {
+    let mut res = vec![command]
+        .into_iter()
+        .chain(args.into_iter())
+        .collect::<Vec<_>>();
+    if let Some(id) = container_id {
+        res.push(id)
+    }
+    res
+}
+
+impl Subcommand {
+    fn build_cmd(&self) -> Vec<String> {
+        use crate::oci::oci_runtime::Subcommand::*;
+        match self {
+            Checkpoint((container_id, args)) => build_vec(
+                self.to_string(),
```

Comment:
this felt bad because this is all repeated code, but I did not think of a rust way around this.